### DISCUSSION TITLE
A dynamic default model attributes feature

### DISCUSSION THREAD

**Discussion by michgeek** [laravel/framework#43976] [Ideas]
We can define default model attributes by declaring them with

```php
class MyModel extends EloquentModel {
    public $attributes = [
        'my_column' => 'default value',
    ];
}
```

It could be useful to have a dynamic default value by declaring an invokable class that will return a value based on custom logic : 

```php
class MyModel extends EloquentModel {
    public $attributes = [
        'my_column' => MyDefaultDynamicValue::class
    ];
}

class MyDefaultDynamicValue implements DefaultModelValue {
    public function __invoke() {
       return auth()->check() ? 'this is a default when I am authenticated' : 'guest default value';
    }
}
```

What do you think ?

**Comment by henzeb** [comment]
Not sure, but couldn't you use Attribute mutators for that?

**Reply by michgeek** [reply]
I believe if you use the mutator it will only be called on mutation, ´MyModel::create()´ will not call the mutator.

**Reply by HeavyMedaI** [reply]
You can use event listener I think.

**Reply by henzeb** [reply]
Or place it in the constructor. Or place it in the boot method.

**Comment by antonkomarev** [comment]
@michgeek I proposed similar idea, but implementation is different, take a look if it fits your requirements:
- https://github.com/laravel/framework/discussions/50436
